File: core/runtime.php/prelude.php
<?php
/* This is free and unencumbered software released into the public domain. */

declare(strict_types=1);

namespace dry;

if (version_compare(phpversion(), '7.0', '<')) {
  die("DRYlib for PHP requires PHP 7.0+\n");
}

////////////////////////////////////////////////////////////////////////////////

function bool($value): bool {
  return boolval($value);
}

function char(int $value): int {
  return intval($value);
}

function complex(Real $real, Real $imaginary): Complex {
  return new Complex($real, $imaginary);
}

function float(float $value): float {
  return floatval($value);
}

function float32(float $value): float {
  return floatval($value);
}

function float64(float $value): float {
  return floatval($value);
}

function int(int $value): int {
  return intval($value);
}

function int8(int $value): int {
  return intval($value);
}

function int16(int $value): int {
  return intval($value);
}

function int32(int $value): int {
  return intval($value);
}

function int64(int $value): int {
  return intval($value);
}

function int128(int $value): int {
  return intval($value);
}

function integer(int $value): Integer {
  return new Integer($value);
}

function natural(int $value): Integer {
  return integer($value);
}

function rational(Integer $numerator, Integer $denominator): Rational {
  return new Rational($numerator, $denominator);
}

function real(float $value): Real {
  return new Real($value);
}

function word(int $value): int {
  return intval($value);
}

function word8(int $value): int {
  return intval($value);
}

function word16(int $value): int {
  return intval($value);
}

function word32(int $value): int {
  return intval($value);
}

function word64(int $value): int {
  return intval($value);
}

////////////////////////////////////////////////////////////////////////////////

interface Value {}

////////////////////////////////////////////////////////////////////////////////

interface Number extends Value {}

////////////////////////////////////////////////////////////////////////////////

class Integer implements Number {
  public $value;

  function __construct(int $value) {
    $this->value = $value;
  }
}

////////////////////////////////////////////////////////////////////////////////

class Rational implements Number {
  public $numerator;
  public $denominator;

  function __construct(Integer $numerator, Integer $denominator) {
    $this->numerator = $numerator;
    $this->denominator = $denominator;
  }
}

////////////////////////////////////////////////////////////////////////////////

class Real implements Number {
  public $value;

  function __construct(float $value) {
    $this->value = $value;
  }
}

////////////////////////////////////////////////////////////////////////////////

class Complex implements Number {
  public $real;
  public $imaginary;

  function __construct(Real $real, Real $imaginary) {
    $this->real = $real;
    $this->imaginary = $imaginary;
  }
}

File: core/runtime.rs/prelude.rs
/* This is free and unencumbered software released into the public domain. */

////////////////////////////////////////////////////////////////////////////////
// Type Definitions

// Boolean (true or false)
type Bool = bool;

// Character
type Char = char;

// Complex number (arbitrary size)
struct Complex {
  real:      Real,
  imaginary: Real,
}

// Floating-point number (native size)
type Float = Float64;

// Floating-point number (32-bit single-precision)
type Float32 = f32;

// Floating-point number (64-bit double-precision)
type Float64 = f64;

// Integer number (native size)
type Int = isize;

// Integer number (8-bit)
type Int8 = i8;

// Integer number (16-bit)
type Int16 = i16;

// Integer number (32-bit)
type Int32 = i32;

// Integer number (64-bit)
type Int64 = i64;

// Integer number (128-bit)
type Int128 = Integer;

// Integer number (arbitrary size)
struct Integer {
  value: Int64 // FIXME
}

// Natural number (arbitrary size)
type Natural = Integer;

// Rational number (arbitrary size)
struct Rational {
  numerator:   Integer,
  denominator: Integer,
}

// Real number (arbitrary size)
struct Real {
  value: Float64 // FIXME
}

// Machine word (native size)
type Word = usize;

// Machine word (8-bit)
type Word8 = u8;

// Machine word (16-bit)
type Word16 = u16;

// Machine word (32-bit)
type Word32 = u32;

// Machine word (64-bit)
type Word64 = u64;

fn main() {} // rustc -A dead_code prelude.rs
